### PR TITLE
Axis domain by data key selector

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

There is a ton to unwrap here in the axis area but it's looking possible.

## Related Issue

https://github.com/recharts/recharts/issues/4583

## Motivation and Context

Remove direct DOM access

## How Has This Been Tested?

npm test

the new hooks and selectors are not yet used anywhere outside of unit tests so I am not running storybook diff

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes

### CODE REVIEW

File: src/state/axisMapSlice.ts [hunk 1]
```diff
@@ -12,7 +13,18 @@ export type AxisId = string | number;
...
+  padding: XAxisPadding;
+  domain: AxisDomain | undefined;
+  allowDataOverflow: boolean;
+  /**
+   * The axis functionality is severely restricted without a dataKey
+   * - but there is still something left, and the prop is optional
+   * so this can also be undefined even in real charts.
+   * There are no defaults.
+   */
+  dataKey: DataKey<any> | undefined;
```
Comment (ckifer) [inline src/state/axisMapSlice.ts:26]: Don't axes grab onto specific dataKeys under certain conditions? Like if you specific on a graphical item then Y axis is defaulted. Vice versa for X axis if layout is swapped

File: src/state/axisSelectors.ts [hunk 2]
```diff
@@ -16,6 +17,62 @@ const unknownScale: ParsedScaleReturn = {
...
+      return undefined;
+    }
+    // Array.prototype.flat has good support everywhere except IE11: https://caniuse.com/?search=flat
+    const allDataSquished = graphicalItemsData
+      .flat(1)
+      .concat(chartData.chartData)
+      .map(entry => getValueByDataKey(entry, axisSettings.dataKey));
+
+    if (axisSettings.type === 'number') {
+      const onlyNumbers = allDataSquished.map(Number).filter(n => Number.isNaN(n) === false);
```
Comment (ckifer) [inline src/state/axisSelectors.ts:66]: Doesn't isNaN only check for the actual value NaN? I think this filter would include strings, etc?

File: src/state/graphicalItemsSlice.ts [hunk 3]
```diff
@@ -24,9 +34,15 @@ const graphicalItemsSlice = createSlice({
     removeBar(state) {
       state.countOfBars -= 1;
     },
+    addGraphicalItemData(state, action: PayloadAction<ChartData>) {
+      state.graphicalItemData.push(action.payload);
+    },
+    removeGraphicalItemData(state, action: PayloadAction<ChartData>) {
+      state.graphicalItemData = state.graphicalItemData.filter(item => item !== action.payload);
```
Comment (ckifer) [inline src/state/graphicalItemsSlice.ts:41]: Is this comparing an entire array?

File: test/state/axisSelectors.spec.tsx [hunk 4]
```diff
@@ -64,3 +64,249 @@ describe('selectAxisScale', () => {
...
+      const spy = vi.fn();
+      const Comp = (): null => {
+        const result = useAppSelector(state => selectDomainOfDataByKey(state, 'xAxis', '0'));
+        spy(result);
+        return null;
+      };
+      render(
+        <BarChart data={PageData} width={100} height={100}>
+          <XAxis dataKey="uv" type="number" />
+          <Customized component={Comp} />
```
Comment (ckifer) [inline test/state/axisSelectors.spec.tsx:113]: Won't these all fail later? Why can't we use renderHook?

**Review by ckifer — APPROVED** [review]
Nice!! Some questions

File: src/state/axisMapSlice.ts (hunk 1 above)
Comment (PavelVanecek) [reply src/state/axisMapSlice.ts:26]: Indeed, graphical items have their own axis maps. That comes next

File: src/state/graphicalItemsSlice.ts (hunk 3 above)
Comment (PavelVanecek) [reply src/state/graphicalItemsSlice.ts:41]: It is, on purpose.

File: src/state/axisSelectors.ts (hunk 2 above)
Comment (PavelVanecek) [reply src/state/axisSelectors.ts:66]: I will add a test

File: test/state/axisSelectors.spec.tsx (hunk 4 above)
Comment (PavelVanecek) [reply test/state/axisSelectors.spec.tsx:113]: Yes all of these will fail. They will be trivial to replace however.

**Review by ckifer — APPROVED** [review]

**Pull request merged**